### DISCUSSION TITLE
#if SKIA_GTK/WPF support in main project

### DISCUSSION THREAD

**Discussion by DmitryBorodiy** [unoplatform/uno#16254] [Q&A]
### What do you need help with?

Previously, I used #if to write platform-specific code for a specific platform, but now this feature has been partially removed. How can I, for example, now call specific APIs of WPF/GTK projects or work with their controls?

**Comment by jeromelaban** [comment]
Thanks for the answer. What do you mean by partially removed?

**Reply by DmitryBorodiy** [reply]
In older Xamarin/UWP project types I could use #if to detect an application running on Linux and even call Linux-specific code inside a #if code block directly from the general project, in new project types I cannot directly access the ability to create platform-specific code, I I will be forced to divide it into different project heads. Moreover, I can no longer directly manage WPF/GTK host windows for finer customization.

**Reply by DmitryBorodiy** [reply]
Where is #if HAS_UNO_SKIA_GTK/WPF ?

**Reply by DmitryBorodiy** [reply]
I have previously experimented with Uno with creating an AllowTransparency window in WPF using Uno. WPF used to have a ContentControl that rendered a page from Uno. I could do whatever I wanted with the window directly from the WPF project. But now you have an AppHead file that complicates this practice, and the WindowHelper simply returns an object that is difficult to bring to a specific window, for example Gtk.Window or System.Windows.Window in the main project, since there is no way to use specific things there at all.

**Reply by lindexi** [reply]
@DmitryBorodiy I write the WPF special code in XxxUno.Skia.Wpf project.

**Reply by DmitryBorodiy** [reply]
Yes, it's possible, but what if I need UI Uno to interact with specific things from WPF?

**Reply by DmitryBorodiy** [reply]
What if you need to use a class that you created in a WPF project from the Uno, what if you need to open another Wpf window but without the Uno host, or create a custom dragable region for the native window WPF / GTK? In the old types of projects, it was possible to fully implement the native experience in those things that Uno does not support.

**Reply by DmitryBorodiy** [reply]
I understand that this can be written separately for each Project head, but it's terribly inconvenient, why don't they make it possible to write platform-specific code directly from the main project, because this way the Uno interface will be able to directly interact with native things from WPF / GTK. While this can still be done on Android, iOS using #if, but on WPF, Gtk is no longer possible, because they updated all project templates with AppHead using.

**Reply by jeromelaban** [reply]
Technically, using shared code in a single project it is not possible to differentiate WPF and Gtk between targets frameworks, because they use the same target framework. Using Gtk and Wpf separate heads, the only way to do this at this time is to use dependency injection and abstractions behind `object` typed instances.

Note that with the Wpf/Gtk, the airgap handling will change in the future depending on the underlying platform.

**Reply by jeromelaban** [reply]
An update on this thread: With the introduction of Uno 5.2, differentiating for each underlying target is changing with the removal of Gtk by default. It is a scenario we'll be working on in a near future, to provide the ability to determine which is the active target in order for native libraries to integrate properly.

**Reply by DmitryBorodiy** [reply]
So, Is gtk still supported by Uno?

**Reply by jeromelaban** [reply]
We will not be removing support for it, but we're not going to make improvements to the GTK support in the future. We'll be moving to native X11/WPF/macOS support directly, and the airspace support will improve soon around these targets.

**Reply by lindexi** [reply]
After conducting tests in Avalonia, I found that using the GTK framework as the underlying layer results in subpar performance. This is due to the fact that the GTK layer informs X11 to render the entire screen for each frame, leading to potential performance issues. However, if X11 is used as the underlying layer, performance can be improved by controlling the rendering range of X11 through XPutImage. The bad news is that I don't seem to have found a mechanism in UNO to refresh only the dirty range.

**Reply by MartinZikmund** [reply]
We will be improving rendering performance as we go, dirty rect rendering is something we would like to introduce as part of it, but there are other things to tackle first (e.g. separate compositor thread)
